File: code/337.house-robber-iii.java
/*
 * @lc app=leetcode id=337 lang=java
 *
 * [337] House Robber III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rob(TreeNode root) {
        int[] res = robTree(root);
        return Math.max(res[0], res[1]);
    }

    public int[] robTree(TreeNode cur) {
        int[] res = new int[2];
        if (cur == null) return res;

        int[] left = robTree(cur.left);
        int[] right = robTree(cur.right);
        res[0] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
        res[1] = cur.val + left[0] + right[0];
        return res;
    }
}
// @lc code=end

File: code/337.house-robber-iii.go
/*
 * @lc app=leetcode id=337 lang=golang
 *
 * [337] House Robber III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rob(root *TreeNode) int {
	res := robTree(root)
	return slices.Max(res)
}

func robTree(node *TreeNode) []int {
	if node == nil {
		return []int{0, 0}
	}
	left := robTree(node.Left)
	right := robTree(node.Right)

	unRobNode := slices.Max(left) + slices.Max(right)
	robNode := node.Val + left[0] + right[0]

	return []int{unRobNode, robNode}
}

// @lc code=end

